src/main/java/org/kohsuke/github/GitHubRequest.java

```diff
@@ -0,0 +1,651 @@
+
+import static java.util.Arrays.asList;
+
+/**
+ * Class {@link GitHubRequest} represents an immutable instance used by the client to determine what information to
+ * retrieve from a GitHub server. Use the {@link Builder} construct a {@link GitHubRequest}.
+ * <p>
+ * NOTE: {@link GitHubRequest} should include the data type to be returned. Any use cases where the same request should
+ * be used to return different types of data could be handled in some other way. However, the return type is currently
+ * not specified until late in the building process, so this is still untyped.
+ * </p>
+ */
+class GitHubRequest {
+
+    private static final List<String> METHODS_WITHOUT_BODY = asList("GET", "DELETE");
+    private final List<Entry> args;
+    private final Map<String, String> headers;
+    private final String apiUrl;
+    private final String urlPath;
+    private final String method;
+    private final InputStream body;
+    private final boolean forceBody;
+
+    private final URL url;
+
+    private GitHubRequest(@Nonnull List<Entry> args,
+            @Nonnull Map<String, String> headers,
+            @Nonnull String apiUrl,
+            @Nonnull String urlPath,
+            @Nonnull String method,
+            @CheckForNull InputStream body,
+            boolean forceBody) throws MalformedURLException {
+        this.args = Collections.unmodifiableList(new ArrayList<>(args));
+        this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
+        this.apiUrl = apiUrl;
+        this.urlPath = urlPath;
+        this.method = method;
+        this.body = body;
+        this.forceBody = forceBody;
+        String tailApiUrl = buildTailApiUrl();
+        url = getApiURL(apiUrl, tailApiUrl);
+    }
+
+    /**
+     * Create a new {@link Builder}.
+     * 
+     * @return a new {@link Builder}.
+     */
+    public static Builder<?> newBuilder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Gets the final GitHub API URL.
+     */
+    @Nonnull
+    static URL getApiURL(String apiUrl, String tailApiUrl) throws MalformedURLException {
+        if (tailApiUrl.startsWith("/")) {
+            if ("github.com".equals(apiUrl)) {// backward compatibility
+                return new URL(GitHubClient.GITHUB_URL + tailApiUrl);
+            } else {
+                return new URL(apiUrl + tailApiUrl);
+            }
+        } else {
+            return new URL(tailApiUrl);
+        }
+    }
+
+    /**
+     * Transform Java Enum into Github constants given its conventions
+     *
+     * @param en
+     *            Enum to be transformed
+     * @return a String containing the value of a Github constant
+     */
+    static String transformEnum(Enum<?> en) {
+        // by convention Java constant names are upper cases, but github uses
+        // lower-case constants. GitHub also uses '-', which in Java we always
+        // replace with '_'
+        return en.toString().toLowerCase(Locale.ENGLISH).replace('_', '-');
```

Comment:
Comments like this are super helpful. :)

Are there other places where we might need to do such a conversion? I only see one instance of `replace('_', '-')` in the whole plugin. Might be fine, it just surprised me that there aren't more.

Reply:
The other places are mostly done automatically by Jackson.   

https://github.com/github-api/github-api/pull/697/files#diff-04099021093ec34af9d55a855cc1e6fdR85-R90

Hm, the `toLowerCase` may not be needed due to line 88.  And line 89 means that, given a field like `create_date`, Jackson will automatically look for a field with the name `createdDate` and then fall back to `create_date` if it doesn't find one. 

However, this setting was only added recently, there are still a lot private fields with underscores in them.

Reply:
Opened #700 to clean this up.